sphinx/search/__init__.py

```diff
@@ -387,12 +393,68 @@ def get_terms(self, fn2index: dict[str, int]) -> tuple[dict[str, list[int] | int
         return rvs
 
+    @staticmethod
+    def _terms_ngrams(terms: dict[str, Any]) -> dict[str, list[str]]:
+        """Extract unique ngrams (currently, trigrams) from the input search terms."""
+        ngrams: dict[str, list[str]] = defaultdict(list)
+        for term in terms:
+            if len(term) >= 3:
+                for i in range(len(term) - 2):
+                    ngrams[term[i:i + 3]].append(term)
+        return {ngram: sorted(set(terms)) for ngram, terms in ngrams.items()}
+
+    @classmethod
+    def _ngrams_trie(cls, ngrams: dict[str, list[str]]) -> _TrieNode:
+        """
+        Given a dictionary of terms and a mapping of ngrams to those terms, compress the
+        mapping into a trie, using an empty-string key to notate the contents (terms
+        containing the ngram) at each node.
+        """
+        root: IndexBuilder._TrieNode = {}
+        for ngram, terms in ngrams.items():
+            location = root
+            for char in ngram:
+                location = location.setdefault(char, {})  # type: ignore[assignment]
+            location[cls._TRIE_CONTENTS] = terms
+        return root
+
+    @classmethod
+    def _minify_trie(cls, trie: _TrieNode, term_offsets: dict[str, int]) -> _TrieNode:
+        """Minify the representation of an ngram-to-terms trie datastructure."""
+        for key, subnode in trie.copy().items():
+            if not isinstance(subnode, dict):
+                continue
+
+            # Replace the string values of terms with their numeric termlist offset.
+            if node_terms := subnode.get(cls._TRIE_CONTENTS, []):
+                offsets = [term_offsets[term] for term in node_terms]
+                # Replace single-valued contents with integers instead of list values.
+                subnode[cls._TRIE_CONTENTS] = offsets[0] if len(offsets) == 1 else offsets  # type: ignore[assignment]
+
+            # Replace single-leaf content nodes (dictionaries) with their (list) contents.
+            if len(subnode) == 1 and cls._TRIE_CONTENTS in subnode:
+                trie[key] = subnode[cls._TRIE_CONTENTS]
+                continue
+
+            # Abbreviate unambiguous paths; (a.empty -> b.empty -> c) becomes (abc).
+            trie[key] = IndexBuilder._minify_trie(subnode, term_offsets)
+            if len(trie[key]) == 1 and cls._TRIE_CONTENTS not in trie[key]:
+                subkey, subnode = trie.pop(key).popitem()  # type: ignore[union-attr]
+                trie[key + subkey] = subnode
+
+        return trie
+
     def freeze(self) -> dict[str, Any]:
         """Create a usable data structure for serializing."""
         docnames, titles = zip(*sorted(self._titles.items()))
         filenames = [self._filenames.get(docname) for docname in docnames]
         fn2index = {f: i for (i, f) in enumerate(docnames)}
         terms, title_terms = self.get_terms(fn2index)
+        term_offsets = {term: idx for idx, term in enumerate(sorted(terms))}
```

Comment:
Note: if we could pre-sort the `terms` keys then we would be able to remove this `sorted(...)` call and also the `sort_keys=True` option when we're writing the index, I think - I'm not 100% certain about this yet though.

Reply:
> Note: if we could pre-sort the `terms` keys then we would be able to remove this `sorted(...)` call and also the `sort_keys=True` option when we're writing the index, I think - I'm not 100% certain about this yet though.

This particular further optimization could and should be handled in a separate, isolated changeset.